### PR TITLE
Add 2D windowing and unwindowing functionality with tests

### PR DESCRIPTION
**Pull request by dnxjay**
- Implemented `apply_2d_windowing` function to apply a specified 2D window (Hann, Hamming, Gaussian)
  - Generates a 2D window by taking the outer product of a 1D window array
  - Multiplies the 2D data by the window, focusing on frequency of interest and reducing edge artifacts
- Implemented `remove_2d_windowing` function to reverse the windowing effect
  - Adds a small epsilon to prevent division by zero during unwindowing
- Added parameterized tests for `apply_2d_windowing` to validate function with multiple window types
- Added test for `remove_2d_windowing` to ensure accurate restoration of original data
  - Comparison focuses on central region to reduce edge effects from windowing
  - Includes tolerance to handle minor discrepancies post-unwindowing
Linting
Linting Status: All code passes linting checks
Contents
Description: This pull request adds the functionality to apply and remove 2D windowing for image data, including Hann, Hamming, and Gaussian window types. It focuses on applying windowing to emphasize specific frequencies, removing windowing while preserving the original data structure, and includes tolerance-based testing to ensure minimal edge artifact interference.
Files Changed:
preparation.py: Added apply_2d_windowing and remove_2d_windowing functions.
test_preparation.py: Added parameterized tests for windowing and unwindowing, focusing on central region data comparison.
Assignment Description:
Assignment: Implement and test 2D windowing/unwindowing for periodic structure data in image processing.

### CODE REVIEW

**Comment by kylemasc917** [comment]
@dhamalakamal Ive added you as a reviewer for this request please do so as soon as possible.

**Review by dhamalakamal — CHANGES_REQUESTED** [review]
Please check the specific pylint errors from PR build and fix them. I have commented on each respective lines as well which requires changes to fix those issues.

**Comment by kylemasc917** [comment]
@dnxjay There are still a number of pylint errors that need to be resolved for this to be merged. Please do so as soon as possible.

**Comment by dnxjay** [comment]
> @dnxjay There are still a number of pylint errors that need to be resolved for this to be merged. Please do so as soon as possible.

i apologize, i thought it was approved, looked at the wrong pylint test but ill work on it right now

**Comment by dnxjay** [comment]
ive been working at the linting for a while but cant get any progress, any recommendations @laserlab

**Comment by laserlab** [comment]
> ive been working at the linting for a while but cant get any progress, any recommendations @laserlab

It’s these lines:  `pickle.dump(sample_data, file)1st
listing is rightfully complaining about changes of outer scope. That means the variable is not given to the function via argument but sample_data is accessed and changed from outside the function which will have unintended consequences.

**Comment by dnxjay** [comment]
> > ive been working at the linting for a while but cant get any progress, any recommendations @laserlab
> 
> 
> 
> It’s these lines:  `pickle.dump(sample_data, file)`
> 
> linting is rightfully complaining about changes of outer scope. That means the variable is not given to the function via argument but sample_data is locally accessed and changed which is in this case the pylint fixture. So probably not the intended idea here. 

How would i go about fixing that? Was that one specifically with the task I'm working on?

**Comment by laserlab** [comment]
> > > ive been working at the linting for a while but cant get any progress, any recommendations @laserlab
> > 
> > 
> > It’s these lines:  `pickle.dump(sample_data, file)`
> > linting is rightfully complaining about changes of outer scope. That means the variable is not given to the function via argument but sample_data is locally accessed and changed which is in this case the pylint fixture. So probably not the intended idea here.
> 
> How would i go about fixing that? Was that one specifically with the task I'm working on?

I do not know if that was in your task, the linting issues should have been there before though if it wasn’t. Maybe @kylemasc917 knows. Fixing is easy though. either using the actual fixture or if that is not what is intended renaming the variable to a unique name.

File: test_preparation.py [hunk 1]
```diff
@@ -257,3 +256,61 @@ def test_load_pickle_data_with_nonexistent_file(capsys):
 
     # Ensure that the error message is printed
     assert "Error loading data: The file '{pickle_file_path}' was not found." in captured.out
+
+# New tests for windowing/unwindowing functions
+import numpy as np
+from preparation import apply_2d_windowing, remove_2d_windowing
+
+@pytest.fixture
+def sample_data():
```
[the remark this answers is not in the document]
Comment (laserlab) [reply test_preparation.py:265]: @dnxjay So that’s the same issue as you asked with a useful answer

**Comment by laserlab** [comment]
@dnxjay also check out the reviewers comment on the same issue

**Comment by dnxjay** [comment]
im not sure how this got merged by myself, are we able undo so? I fixed the pylint issues and when i was syncing my fork something went wrong. The linting issues should be fixed now.

**Pull request closed without merging**